internal/mailer/mailme.go

```diff
@@ -0,0 +1,193 @@
+package mailer
+
+import (
+	"bytes"
+	"errors"
+	"html/template"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"gopkg.in/gomail.v2"
+
+	"github.com/sirupsen/logrus"
+)
+
+// TemplateRetries is the amount of time MailMe will try to fetch a URL before giving up
+const TemplateRetries = 3
+
+// TemplateExpiration is the time period that the template will be cached for
+const TemplateExpiration = 10 * time.Second
+
+// MailmeMailer lets MailMe send templated mails
+type MailmeMailer struct {
+	From      string
+	Host      string
+	Port      int
+	User      string
+	Pass      string
+	BaseURL   string
+	LocalName string
+	FuncMap   template.FuncMap
+	cache     *TemplateCache
+	Logger    logrus.FieldLogger
+}
+
+// Mail sends a templated mail. It will try to load the template from a URL, and
+// otherwise fall back to the default
+func (m *MailmeMailer) Mail(to, subjectTemplate, templateURL, defaultTemplate string, templateData map[string]interface{}) error {
+	if m.FuncMap == nil {
+		m.FuncMap = map[string]interface{}{}
+	}
+	if m.cache == nil {
+		m.cache = &TemplateCache{
+			templates: map[string]*MailTemplate{},
+			funcMap:   m.FuncMap,
+			logger:    m.Logger,
+		}
+	}
+
+	tmp, err := template.New("Subject").Funcs(template.FuncMap(m.FuncMap)).Parse(subjectTemplate)
+	if err != nil {
+		return err
+	}
+
+	subject := &bytes.Buffer{}
+	err = tmp.Execute(subject, templateData)
+	if err != nil {
+		return err
+	}
+	body, err := m.MailBody(templateURL, defaultTemplate, templateData)
+	if err != nil {
+		return err
+	}
+
+	mail := gomail.NewMessage()
+	mail.SetHeader("From", m.From)
+	mail.SetHeader("To", to)
+	mail.SetHeader("Subject", subject.String())
+	mail.SetBody("text/html", body)
+
+	dial := gomail.NewDialer(m.Host, m.Port, m.User, m.Pass)
+	if m.LocalName != "" {
+		dial.LocalName = m.LocalName
+	}
+	return dial.DialAndSend(mail)
+
+}
+
+type MailTemplate struct {
+	tmp       *template.Template
+	expiresAt time.Time
+}
+
+type TemplateCache struct {
+	templates map[string]*MailTemplate
+	mutex     sync.Mutex
+	funcMap   template.FuncMap
+	logger    logrus.FieldLogger
+}
+
+func (t *TemplateCache) Get(url string) (*template.Template, error) {
+	cached, ok := t.templates[url]
```

Comment:
This looks like a race, we read from t.templates without mutex here, but write to it in Set() with one. Easy fix though.